bluemira/equilibria/coils.py

```diff
@@ -85,634 +108,631 @@ def generate_name(coil: Any, idx: Optional[int] = None):
+            CoilNumber.__no_counter += 1
+        elif ctype == CoilType.CS:
+            idx = CoilNumber.__CS_counter
+            CoilNumber.__CS_counter += 1
+        elif ctype == CoilType.PF:
+            idx = CoilNumber.__PF_counter
+            CoilNumber.__PF_counter += 1
 
+        return idx
 
-def get_max_current(dx, dz, j_max):
-    """
-    Get the maximum current in a coil cross-sectional area
-
-    Parameters
-    ----------
-    dx: float
-        Coil half-width [m]
-    dz: float
-        Coil half-height [m]
-    j_max: float
-        Coil current density [MA/m^2]
 
-    Returns
-    -------
-    max_current: float
-        Maximum current [A]
+def _sum_all(func: Optional[callable] = None, *, axis: int = 0) -> callable:
     """
-    return abs(j_max * 1e6 * (4 * dx * dz))
-
-
-class CoilSizer:
+    Sum all outputs of a function on a given axis
     """
-    Coil sizing utility class (observer pattern).
-
-    Parameters
-    ----------
-    coil: Coil
-        Coil to size
 
-    Notes
-    -----
-    Maximum currents are not enforced anywhere in Coils. If you want constrain currents,
-    you should use constrained optimisation techniques (with current bounds).
-    """
+    def wrapper(*args, **kwargs):
+        ret = func(*args, **kwargs)
+        return ret.sum(axis=axis)
 
-    def __init__(self, coil):
-        self.update(coil)
+    if func is None:
 
-        dxdz_specified = is_num(self.dx) and is_num(self.dz)
+        def decorator(func):
+            return update_wrapper(wrapper, func)
 
-        if not dxdz_specified and not (self.dx is None and self.dz is None):
-            # Check that we don't have dx = None and dz = float or vice versa
-            raise EquilibriaError("Must specify either dx and dz or neither.")
+        return decorator
+    else:
+        return update_wrapper(wrapper, func)
 
-        if dxdz_specified and not self.flag_sizefix:
-            # If dx and dz are specified, we presume the coil size should remain fixed
-            self.flag_sizefix = True
 
-        if dxdz_specified:
-            self._set_coil_attributes(coil)
+class CoilFieldsMixin:
 
-        if not dxdz_specified and not self.j_max:
-            # Check there is a viable way to size the coil
-            raise EquilibriaError("Must specify either dx and dz or j_max.")
+    __slots__ = (
+        "_quad_dx",
+        "_quad_dz",
+        "_quad_x",
+        "_quad_z",
+        "_quad_weighting",
+        "_einsum_str",
+    )
 
-        if not dxdz_specified and self.flag_sizefix:
-            # If dx and dz are not specified, we cannot fix the size of the coil
-            self.flag_sizefix = False
+    def __init__(self, weighting=None):
+        if type(self) == CoilFieldsMixin:
+            raise TypeError("Can't be initialised directly")
 
-        coil.flag_sizefix = self.flag_sizefix
+        self.discretise(weighting)
 
-    def __call__(self, coil, current=None):
+    def _pad_discretisation(
```

Comment:
I don't understand why we need to pad at all... why is this?

Reply:
The padding is so we can use the qudrature properties in one array even though they may be of different lengthed vectors due to coils eg having the same discretisation but are different sizes. No extra calculations are done on these zeros as of [17af92d](https://github.com/Fusion-Power-Plant-Framework/bluemira/pull/708/commits/17af92dbf3be1c959aff6ae6e41b41adfbb9c9a6). The user also is unlikely to interact with them.